### ISSUE TITLE
"Downloaded language server binary is not funtional" error on Windows

### ISSUE THREAD

**Issue by yoshuawuyts** [rust-analyzer/rust-analyzer#3087]
After trying to install the language server binary through the popup today, the following error occurred:

<img width="960" alt="Screenshot 2020-02-10 17 43 21" src="https://user-images.githubusercontent.com/2467194/74170248-ea378300-4c2c-11ea-8869-282609626ce8.png">

This occurs on Rust 1.41 on Windows 10. Possibly related to https://github.com/rust-analyzer/rust-analyzer/issues/3073

**Comment by Veetaha** [comment]
Can you please open command palette with <knd>Ctrl+Shift+P</kbd>, write `Toggle Developer Tools` and press enter to open the Developer Tools, then open Console tab and send its contents here?
Also please give the contents of Help > About window in VSCode?

**Comment by yoshuawuyts** [comment]
@Veetaha I can't seem to get the popup to appear again. Is there a recommended way to make that happen?

Actually I'm suspecting it may have been a one-off glitch because now that I've refreshed it seems like it's working?

---

Console output:

```txt
[Intervention] Slow network is detected. See https://www.chromestatus.com/feature/5636954674692096 for more details. Fallback font will be used while loading: file:///C:/Users/yoshu/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/base/browser/ui/codiconLabel/codicon/codicon.ttf?c4e66586cd3ad4acc55fc456c0760dec
actionbar.ts:669 [Intervention] Slow network is detected. See https://www.chromestatus.com/feature/5636954674692096 for more details. Fallback font will be used while loading: file:///C:/Users/yoshu/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/extensions/theme-seti/icons/seti.woff
2console.ts:137 [Extension Host] (node:1696) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
t.log @ console.ts:137
$logExtensionHostMessage @ mainThreadConsole.ts:39
_doInvokeHandler @ rpcProtocol.ts:398
_invokeHandler @ rpcProtocol.ts:383
_receiveRequest @ rpcProtocol.ts:299
_receiveOneMessage @ rpcProtocol.ts:226
(anonymous) @ rpcProtocol.ts:101
fire @ event.ts:581
fire @ ipc.net.ts:453
_receiveMessage @ ipc.net.ts:733
(anonymous) @ ipc.net.ts:592
fire @ event.ts:581
acceptChunk @ ipc.net.ts:239
(anonymous) @ ipc.net.ts:200
t @ ipc.net.ts:28
emit @ events.js:200
addChunk @ _stream_readable.js:294
readableAddChunk @ _stream_readable.js:275
Readable.push @ _stream_readable.js:210
onStreamRead @ internal/stream_base_commons.js:166
console.ts:137 [Extension Host] Checked binary availablity via --version {status: 0, signal: null, output: Array(3), pid: 15748, stdout: {…}, …}
console.ts:137 [Extension Host] c:\Users\yoshu\AppData\Roaming\Code\User\globalStorage\matklad.rust-analyzer\ra_lsp_server-windows.exe --version output: (3) ["null", "rust-analyzer 5b703bd↵", ""]
```

About:


```txt
Version: 1.41.1 (user setup)
Commit: 26076a4de974ead31f97692a0d32f90d735645c0
Date: 2019-12-18T14:58:56.166Z
Electron: 6.1.5
Chrome: 76.0.3809.146
Node.js: 12.4.0
V8: 7.6.303.31-electron.0
OS: Windows_NT x64 10.0.18363
```

**Comment by Veetaha** [comment]
@yoshuawuyts  I am sorry to admit that we already stumbled with your bug, the error was `EBUSY`, but it just disappeared away like in your case so we were not able to track the actual reason. Now you do have a working binary and it must not reproduce.

@matklad I think to kill this bug we will need to setup logging to a file from vscode extension

**Comment by Veetaha** [comment]
I also wonder whether this **always** happens on Windows or just with a case-by-base basis...

**Comment by yoshuawuyts** [comment]
@Veetaha Going to try running this again on my Windows desktop to see if I can reproduce it once more.

**Comment by yoshuawuyts** [comment]
This time around it went without a problem:

![Screenshot 2020-02-10 18 58 30](https://user-images.githubusercontent.com/2467194/74176169-777fd500-4c37-11ea-9448-ab311666d3c1.png)
![Screenshot 2020-02-10 18 58 26](https://user-images.githubusercontent.com/2467194/74176171-78186b80-4c37-11ea-807e-b4d49e25bbd1.png)
![Screenshot 2020-02-10 18 58 18](https://user-images.githubusercontent.com/2467194/74176174-79499880-4c37-11ea-94bd-904d9cb5b994.png)

**Comment by matklad** [comment]
I would try to check how other extensions deal with this issue.

**Comment by Veetaha** [comment]
@yoshuawuyts Thank you for the report and efforts, this is very valuable. I see that you have the same symptoms, reinstalling does not reproduce the error.
@matklad sure, will try

**Comment by deltaphc** [comment]
Just passing by to confirm. I had been tracking `master` up until today, and decided to try the marketplace extension on Windows.

I deleted my rust-analyzer repo dir, uninstalled the manually-installed extension, and deleted ra-lsp-server from .cargo/bin, then installed the marketplace extension. It prompted me to download, downloaded it, but then said the binary was 'non-functional' after it was done.

Did RA's "restart language server" command, and it started working fine.

**Comment by matklad** [comment]
Stupid idea, but maybe we are not closing the fs stream properly?

**Comment by Veetaha** [comment]
@matklad, this is ridiculous, but yeah, I independently came to this conclusion in #3090

**Comment by marcogroppo** [comment]
@matklad I think we should explicitly close() the output file stream here, after we receive a `finish` event: https://github.com/rust-analyzer/rust-analyzer/blob/7cba77ed4e4207b2e24b8dd57723368c2717bb2a/editors/code/src/installation/download_file.ts#L32

Something like that:
```javascript
const outFile = fs.createWriteStream(destination);
response.pipe(outFile)
outFile.on('finish', () => outFile.close())
```

**Comment by Veetaha** [comment]
@marcogroppo,
From [pipe docs:](https://nodejs.org/docs/latest-v12.x/api/stream.html#stream_readable_pipe_destination_options)
```
By default, stream.end() is called on the destination Writable stream when the source Readable stream emits 'end', so that the destination is no longer writable.
```

**Comment by marcogroppo** [comment]
@Veetaha does this imply that the underlying file descriptor is *immediately* closed? 🤔 (I don't know)

**Comment by TheAifam5** [comment]
I have similar issue but on Linux. Prompts me to download the `rust-analyzer` binaries, even I have them installed. Do I still need `rust-analyzer` installed globally on my system?

**Comment by Veetaha** [comment]
@TheAifam5 , what do you mean by system-wide? Did you press "Download now" and then it prompted you after to download them one more time after the reboot, or do you mean that you have `ra_lsp_server` in your path. If the latter, please set `"rust-analyzer.raLspServerPath": "ra_lsp_server"`

**Comment by TheAifam5** [comment]
@Veetaha yup, I mean system-wide (in /usr/bin) which is accessible through PATH. Thank you for the advice.
